aten/src/ATen/native/cuda/FractionalMaxPool2d.cu

```diff
@@ -0,0 +1,375 @@
+
+  // Each thread generates a specific output point
+  if (ourOutputPoint < output.size(2) * output.size(3)) {
+    int outputW = ourOutputPoint % output.size(3);
+    int outputH = ourOutputPoint / output.size(3);
+
+    int poolW = get_interval<scalar_t, accscalar_t>(
+      static_cast<accscalar_t>(samples[batch][plane][0]),
+        outputW, input.size(3), output.size(3), poolSizeW);
+    int poolH = get_interval<scalar_t, accscalar_t>(
+      static_cast<accscalar_t>(samples[batch][plane][1]),
+        outputH, input.size(2), output.size(2), poolSizeH);
+
+    scalar_t maxVal = at::numeric_limits<scalar_t>::lowest();
+    int maxIndex = -1;
+
+    for (int h = poolH; h < poolH + poolSizeH; ++h) {
+      if (PoolSizeWStatic == -1) {
+        for (int w = poolW; w < poolW + poolSizeW; ++w) {
+          scalar_t val = input[batch][plane][h][w];
+          // for consistency with THNN, favor the first max
+          if (val > maxVal) {
+            maxIndex = h * input.size(3) + w;
+            maxVal = val;
+          }
+        }
+      } else {
+#pragma unroll
+        for (int i = 0; i < PoolSizeWStatic; ++i) {
+          int w = i + poolW;
+          scalar_t val = input[batch][plane][h][w];
+          // for consistency with THNN, favor the first max
+          if (val > maxVal) {
+            maxIndex = h * input.size(3) + w;
+            maxVal = val;
+          }
+        }
+      }
+    }
+
+    assert(maxVal != at::numeric_limits<scalar_t>::lowest());
+    assert(maxIndex != -1);
+
+    indices[batch][plane][outputH][outputW] = maxIndex;
+    output[batch][plane][outputH][outputW] = maxVal;
+  }
+}
+
+template <typename scalar_t>
+__global__ void fractional_max_pool2d_backward_out_cuda_frame(
+  PackedTensorAccessor<scalar_t, 4> gradInput,
+  PackedTensorAccessor<scalar_t, 4> gradOutput,
+  PackedTensorAccessor<int64_t, 4> indices) {
+  // Output (h, w) point that this thread is responsible for
+  int ourOutputPoint = threadIdx.x + blockIdx.x * blockDim.x;
+  int plane = blockIdx.y;
+  int batch = blockIdx.z;
+
+  // Each thread generates a specific output point
+  if (ourOutputPoint < gradOutput.size(2) *
+    gradOutput.size(3)) {
+    int outputW = ourOutputPoint % gradOutput.size(3);
+    int outputH = ourOutputPoint / gradOutput.size(3);
+
+    int index = indices[batch][plane][outputH][outputW];
+    assert(index >= 0);
+    int inputW = index % gradInput.size(3);
+    int inputH = index / gradInput.size(3);
+    assert(inputH < gradInput.size(2));
+
+    atomicAdd(
+      &gradInput[batch][plane][inputH][inputW],
+      gradOutput[batch][plane][outputH][outputW]
+    );
+  }
+}
+
+void fractional_max_pool2d_out_cuda_template(
+  Tensor & output,
+  Tensor& indices,
+  const Tensor& input,
+  IntList pool_size,
+  IntList output_size,
+  const Tensor& randomSamples) {
+  int planeDim = 0;
+  int dimh = 1;
+  int dimw = 2;
+  int numBatch = 1;
+
+  int ndims = input.ndimension();
+  for (int i = 0; i < ndims; i++) {
```

Comment:
nit: You can just test that numel > 0 to test if something is non-empty

Reply:
done